pkg/controller/hostpathprovisioner/contoller_test.go

```diff
@@ -0,0 +1,716 @@
+/*
+Copyright 2019 The hostpath provisioner operator Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package hostpathprovisioner
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	secv1 "github.com/openshift/api/security/v1"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes/scheme"
+
+	"kubevirt.io/hostpath-provisioner-operator/pkg/apis/hostpathprovisioner/v1alpha1"
+	"kubevirt.io/hostpath-provisioner-operator/version"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	conditions "github.com/openshift/custom-resource-status/conditions/v1"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+const (
+	versionString = "1.0.1"
+)
+
+var _ = Describe("Controller reconcile loop", func() {
+	var (
+		cr *v1alpha1.HostPathProvisioner
+		cl client.Client
+		r  *ReconcileHostPathProvisioner
+	)
+
+	BeforeEach(func() {
+		watchNamespaceFunc = func() (string, error) {
+			return "test-namespace", nil
+		}
+		version.VersionStringFunc = func() *string {
+			value := versionString
+			return &value
+		}
+
+		cr = &v1alpha1.HostPathProvisioner{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "test-name",
+				Namespace: "test-namespace",
+			},
+			Spec: v1alpha1.HostPathProvisionerSpec{
+				ImageRegistry:   "test-registry",
+				ImagePullPolicy: corev1.PullAlways,
+				ImageTag:        "canary",
+				PathConfig: v1alpha1.PathConfig{
+					Path:            "/tmp/test",
```

Comment:
Should we test that a CS with empty path is rejected?

Reply:
We would need a functional test for that, which would require a cluster to be up. I didn't want want to involve clusters at this point.

Reply:
What will the controller do if it encounters a CR with the path unset?

Reply:
It will fail according to this: https://github.com/kubevirt/hostpath-provisioner/blob/master/cmd/provisioner/hostpath-provisioner.go#L74